internal/cli/resolve.go

```diff
@@ -0,0 +1,254 @@
+				[]build.Option{
+					build.WithLogger(logger),
+					build.WithConfig(args[0]),
+					build.WithExtraKeys(extraKeys),
+					build.WithExtraRepos(extraRepos),
+					build.WithDebugLogging(debugEnabled),
+				},
+			)
+		},
+	}
+
+	cmd.Flags().StringSliceVarP(&extraKeys, "keyring-append", "k", []string{}, "path to extra keys to include in the keyring")
+	cmd.Flags().StringSliceVarP(&extraRepos, "repository-append", "r", []string{}, "path to extra repositories to include")
+	cmd.Flags().StringSliceVar(&archstrs, "arch", nil, "architectures to build for (e.g., x86_64,ppc64le,arm64) -- default is all, unless specified in config. Can also use 'host' to indicate arch of host this is running on")
+	cmd.Flags().StringVar(&output, "output", "", "path to file where lock file will be written")
+	cmd.Flags().StringSliceVar(&logPolicy, "log-policy", []string{}, "logging policy to use")
+	cmd.Flags().BoolVar(&debugEnabled, "debug", false, "enable debug logging")
+	cmd.Flags().BoolVar(&quietEnabled, "quiet", false, "disable logging")
+
+	return cmd
+}
+
+func ResolveCmd(ctx context.Context, output string, archs []types.Architecture, opts []build.Option) error {
+	wd, err := os.MkdirTemp("", "apko-*")
+	if err != nil {
+		return fmt.Errorf("failed to create working directory: %w", err)
+	}
+	defer os.RemoveAll(wd)
+
+	o, ic, err := build.NewOptions(opts...)
+	if err != nil {
+		return err
+	}
+
+	// cases:
+	// - archs set: use those archs
+	// - archs not set, bc.ImageConfiguration.Archs set: use Config archs
+	// - archs not set, bc.ImageConfiguration.Archs not set: use all archs
+	switch {
+	case len(archs) != 0:
+		ic.Archs = archs
+	case len(ic.Archs) != 0:
+		// do nothing
+	default:
+		ic.Archs = types.AllArchs
+	}
+	// save the final set we will build
+	archs = ic.Archs
+	o.Logger().Infof("Determining packages for %d architectures: %+v", len(ic.Archs), ic.Archs)
+
+	// The build context options is sometimes copied in the next functions. Ensure
+	// we have the directory defined and created by invoking the function early.
+	defer os.RemoveAll(o.TempDir())
+
+	lock := lock{
+		Version: "v1",
+		Contents: lockContents{
+			Packages:     []lockPkg{},
+			Repositories: []lockRepo{},
+			Keyrings:     []lockKeyring{},
+		},
+	}
+
+	repositories := map[string]bool{}
+
+	for _, keyring := range ic.Contents.Keyring {
+		lock.Contents.Keyrings = append(lock.Contents.Keyrings, lockKeyring{
+			Name: stripUrlScheme(keyring),
+			Url:  keyring,
+		})
+	}
+
+	for _, arch := range archs {
+		arch := arch
+		// working directory for this architecture
+		wd := filepath.Join(wd, arch.ToAPK())
+		bopts := append(slices.Clone(opts), build.WithArch(arch))
+		fs := apkfs.DirFS(wd, apkfs.WithCreateDir())
+		bc, err := build.New(ctx, fs, bopts...)
+		if err != nil {
+			return err
+		}
+
+		resolvedPkgs, err := bc.Resolve(ctx)
+
+		if err != nil {
+			return fmt.Errorf("failed to get package list for image: %w", err)
+		}
+
+		for _, rpkg := range resolvedPkgs {
```

Comment:
I think we want to fail if the archs produce inconsistent resolved configs. That's what we do in tf-apko and it generally helps prevent confusion later.

Reply:
do you have a link I can look at? I am not sure about what to do here.

Reply:
https://github.com/chainguard-dev/terraform-provider-apko/blob/002f3cb0e3d06f356357e0e484224a54bbc667f6/internal/provider/config_data_source.go#L342 is where we align these in tf-apko -- it's pretty gnarly unfortunately.

Basically if two archs of the same image can't agree about what packages are in there (e.g., because arm builds are ahead of x86) then we fail rather than produce an image with inconsistent versions.

We could do the same here, possibly by resolving off x86 and checking that arm is the same. `apko` itself won't complain about this, but maybe it should.

It's not a blocker, but something we might want a TODO about in case this causes confusion in the future.